### DISCUSSION TITLE
Unable to log into R Studio server

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1196] [Q&A] [answered]
I'm running a pretty intensive randomization test on the R Studio server that involves interacting `calculateDiffMeth` 1000 times:

```
for (i in 1:1000) {
  pHRandTreat <- sample(sampleMetadataFem$pHTreatment,
                        size = length(sampleMetadataFem$pHTreatment),
                        replace = FALSE)
  pHRandDML <- methylKit::reorganize(methylationInformationFilteredCov5Fem,
                                     sample.ids = sampleMetadataFem$sampleID,
                                     treatment = pHRandTreat) %>%
    methylKit::calculateDiffMeth(., covariates = covariateMetadataFem,
                                 overdispersion = "MN",
                                 test = "Chisq", mc.cores = 40) %>%
    methylKit::getMethylDiff(., difference = 25, qvalue = 0.01) %>%
    nrow() -> pHRandTest25Fem[i]
}
```

My session timed out after 60 minutes and when I tried to log back in, I can't even access http://localhost:8787/ ! The log in page is not loading, so I can't check on the progress of my code. I killed my mox tunnel and opened a new one to see if I could load http://localhost:8787/ and that didn't work.

Is there anything I can do? My R Studio server SLURM script is still running, but I guess I could end that too and just interrupt whatever is running.

**Comment by kubu4** [comment]
Is your terminal throwing an error? 

Regardless, I suspect that the port 8787 didn't close properly after the session timeout. So, you need to kill the SSH session that's still occupying that port:

`lsof -ti:8787 | xargs kill -9`

Then, you'll need to re-establish your tunnel and then the web browser URL should work (refresh the page).

**Reply by yaaminiv** [reply]
My terminal isn't throwing an error. 

> you need to kill the SSH session that's still occupying that port

I ran that code in my terminal and opened a new SSH session, but I still can't access the log in page (I opened a new window with the URL)

**Reply by kubu4** [reply]
> The log in page is not loading

What does the web page look like/what does it do?

**Reply by yaaminiv** [reply]
![Screen Shot 2021-05-03 at 2 04 42 PM](https://user-images.githubusercontent.com/22335838/116933570-9261af00-ac18-11eb-9954-5828e8caa56c.png)

It's stuck in a state of perpetual loading (I tried to log in again a few minutes ago since I killed my tunnel yesterday)

*edited to include screenshot context*

**Reply by kubu4** [reply]
What's the longest you've waited without closing the browser (tab) and the SSH tunnel?

**Reply by kubu4** [reply]
Basically, I read that the session load time is highly dependent upon the amount of information in the R environment. So, if you have large data objects, it will lead to lengthy session load times. This could be the issue.

**Reply by yaaminiv** [reply]
> What's the longest you've waited without closing the browser (tab) and the SSH tunnel?

Maybe 15 min? I can let this sit for a few hours since I have a lot of data in the R environment!

**Reply by kubu4** [reply]
Any luck with waiting on the web browser?

Another option is to kill the SLURM job entirely and edit the SLURM script. If you add the following to your script (just add it immediately after the line which loads the singularity module), it should prevent session timeouts from occurring.

```shell
# Do not suspend idle sessions.
# Alternative to setting session-timeout-minutes=0 in /etc/rstudio/rsession.conf
# https://github.com/rstudio/rstudio/blob/v1.4.1106/src/cpp/server/ServerSessionManager.cpp#L126
export SINGULARITYENV_RSTUDIO_SESSION_TIMEOUT=0
```

This was taken from this: https://github.com/rocker-org/website/blob/master/content/use/singularity.md

**Reply by yaaminiv** [reply]
<img width="438" alt="Screen Shot 2021-05-04 at 7 57 54 AM" src="https://user-images.githubusercontent.com/22335838/117045725-aebb2580-acc4-11eb-897b-62a7e8a0708b.png">

Still loading, so I just killed my script and will add that code. Thanks!

**Reply by kubu4** [reply]
Dang. Sorry. Hope this helps!!

**Comment by yaaminiv** [comment]
Update: I started a randomization test and left for a few hours, came back and I had been signed out after 60 min of inactivity! I put in my user name and pw, but I can't sign in! When I click "Sign In," nothing happens and I am not directed to my R Studio server session.

![Screen Shot 2021-05-04 at 2 57 31 PM](https://user-images.githubusercontent.com/22335838/117075055-0ec3c300-ace9-11eb-8096-23b39808e440.png)

(ignore the "connection refused" error in the Terminal window, I was trying to tunnel in this morning using information from a previous SLURM script. I set up the tunnel correctly around 12 and haven't had any Terminal errors since)

*edited to include a description of what happens when I try to sign in*

**Reply by kubu4** [reply]
Don't have any real suggestions. When I tested things, I definitely signed in/out of a single session multiple times and didn't have any issues. I'd close the browser (quit it entirely) and quit Terminal entirely. Then, re-establish the tunnel. Then, open web browser and add the appropriate URL for the RStudio Server. Copy/paste username and password found in your SLURM output (make sure you're copying from the correct SLURM output; just in case you have more than one).

Alternatively, kill your SLURM job and modify your SBATCH script so that the part which executes the Singularity container looks like this:

```shell
# This example bind mounts the /gscratch/scrubbed/${USER} directory on the host into the Singularity container.
# By default the only host file systems mounted within the container are $HOME, /tmp, /proc, /sys, and /dev.
singularity exec \
--bind="$TMPDIR/var/lib:/var/lib/rstudio-server" \
--bind="$TMPDIR/var/run:/var/run/rstudio-server" \
--bind="$TMPDIR/tmp:/tmp" \
--bind=/gscratch/scrubbed/${USER} \
${container_path}/${container} \
rserver --www-port ${PORT} --auth-none=0 --auth-pam-helper-path=pam-helper --auth-stay-signed-in-days=30 --auth-timeout-minutes=0
```

There are two changes there:

1. Removes automatic sign out
2. Keeps your session signed in for 30 days.

**Reply by yaaminiv** [reply]
I put in my username/pw and let it load over night to see what happened, and I got this menu:

![Screen Shot 2021-05-05 at 9 21 08 AM](https://user-images.githubusercontent.com/22335838/117175208-6f074300-ad83-11eb-90d3-fa09e7b5547a.png)

Not entirely sure what "Safe Mode" is, but I clicked that to see what would happen. Still loading.

I wonder if this is related to the fact that I'm now using `mc.cores` in my `calculateDiffMeth` code? I had [issues with that previously](https://github.com/RobertsLab/resources/discussions/1148#discussioncomment-473523).

**Comment by yaaminiv** [comment, marked as answer]
> Alternatively, kill your SLURM job and modify your SBATCH script so that the part which executes the Singularity container looks like this

I created a modified script (found [here](https://github.com/RobertsLab/project-oyster-oa/blob/master/code/Haws/04-methylKit-RStudio.sh)). I get an error in my slurm.out:

![Screen Shot 2021-05-07 at 11 23 51 AM](https://user-images.githubusercontent.com/22335838/117492624-b548d780-af26-11eb-8ad2-9b773e7e6528.png)

I'm also unable to tunnel into mox:

![Screen Shot 2021-05-07 at 11 24 15 AM](https://user-images.githubusercontent.com/22335838/117492658-c2fe5d00-af26-11eb-95cc-7786cdf4a72c.png)

Am I missing any code for the singularity container?

**Reply by yaaminiv** [reply]
Was missing `{container}` variable in my script, added container path and it worked!
